lms/static/scripts/frontend_apps/components/SubmitGradeForm.tsx

```diff
@@ -169,6 +226,88 @@ export default function SubmitGradeForm({
+                classes={classnames(
+                  'border border-r-0 rounded-none ring-inset h-full relative',
+                  'disabled:opacity-50'
+                )}
+              >
+                {commentHasBeenEdited && (
+                  <div className="absolute top-[5px] right-[4px] rounded-full p-1 bg-red-error" />
+                )}
+              </Button>
+              {showCommentControls && (
+                <div
+                  className={classnames(
+                    'w-80 p-3',
+                    'shadow border rounded bg-white',
+                    'absolute top-[calc(100%+3px)] right-0'
+                  )}
+                >
+                  <PointerUpIcon
+                    className={classnames(
+                      'text-grey-3 fill-white',
+                      'absolute inline z-2 w-[15px]',
+                      // Position arrow over "Add comment" button
+                      'right-[7px] top-[-9px]'
+                    )}
+                  />
+                  <div className="flex items-center">
+                    <label htmlFor={commentId} className="font-bold">
+                      Add a comment:
+                    </label>
+                    <div className="grow" />
+                    <IconButton
+                      title="Close comment"
+                      icon={CancelIcon}
+                      classes="hover:bg-grey-3/50"
+                      onClick={() => setShowCommentControls(false)}
+                    />
+                  </div>
+                  <Textarea
+                    id={commentId}
+                    feedback={commentHasBeenEdited ? 'warning' : undefined}
+                    classes="mt-1"
+                    rows={10}
+                    value={commentValue}
+                    onChange={e => {
+                      const newComment = (e.target as HTMLTextAreaElement)
+                        .value;
+                      setCommentValue(newComment);
+
+                      if (student) {
+                        studentComments.current.set(student.userid, newComment);
+                      }
+                    }}
+                  />
+                  <div className="flex flex-row-reverse space-x-2 space-x-reverse mt-3">
+                    <Button
+                      variant="primary"
+                      disabled={disabled}
+                      onClick={onSubmitGrade}
+                    >
+                      {commentHasBeenEdited
+                        ? 'Update comment'
+                        : 'Save & Submit'}
+                    </Button>
+                    <Button
+                      icon={CancelIcon}
+                      onClick={() => setShowCommentControls(false)}
+                    >
+                      Cancel
+                    </Button>
+                  </div>
+                </div>
+              )}
+            </span>
+          )}
```

Comment:
I'm trying to extract this to a separated component, as I think it will simplify testing it, but still don't know if it will make things more complex and I may discard it for now.